doc/man/man3/XOM.3.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man3/XOM.3.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2011  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH XOM 3 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Library Calls"
.\"
.\"
.SH NAME
.B XOM
\- 
.\"
.\"
.SH SYNOPSIS
.PP
.B #include <xom.h>
.HP
.BI "OM_return_code om_copy();"
.PD 0
.HP
.BI "OM_return_code om_copy_value();"
.HP
.BI "OM_return_code om_create();"
.HP
.BI "OM_return_code om_decode();"
.HP
.BI "OM_return_code om_delete();"
.HP
.BI "OM_return_code om_encode();"
.HP
.BI "OM_return_code om_get();"
.HP
.BI "OM_return_code om_instance();"
.HP
.BI "OM_return_code om_put();"
.HP
.BI "OM_return_code om_read();"
.HP
.BI "OM_return_code om_remove();"
.HP
.BI "OM_return_code om_write();"
.PD
.\"
.\"
.SH ARGUMENTS
.PP
Various.
.\"
.\"
.de #v
.ad l
.nh
..
.\"
.\"
.SH TYPES
.PP
.TP
.B OM_sint
.IR Sint :
the positive and negative integers representable in 16 bits (2's complement).
.TP
.B OM_sint16
.IR Sint16 :
the positive and negative integers representable in 16 bits (2's complement).
.TP
.B OM_sint32
.IR Sint32 :
the positive and negative integers representable in 32 bits (2's complement).
.TP
.B OM_uint
.IR Uint :
the non-negative integers representable in 16 bits.
.TP
.B OM_uint16
.IR Uint16 :
the non-negative integers representable in 16 bits.
.TP
.B OM_uint32
.IR Uint32 :
the non-negative integers representable in 32 bits.
.TP
.B OM_double
.IR Double :
the positive and negative floating point numbers representable in 64 bits.
.TP
.B OM_boolean
.IR Boolean :
type definition for a Boolean data value.
.TP
.B OM_descriptor
.IR Descriptor :
type definition for describing an attribute type and value.
.TP
.B OM_enumeration
.IR Enumeration :
type definition for an Enumerated data value.
.TP
.B OM_exclusions
.IR Exclusions :
type definition for 'exclusions' argument for the Get function.
.TP
.B OM_integer
.IR Integer :
type definition for an Integer data value.
.TP
.B OM_modification
.IR Modifications :
type definition for 'modifications' argument for the Put function.
.TP
.B OM_object
.IR Object :
type definition for a handle to either a private or a public object.
.TP
.B OM_object_identifier
.IR Object\0Identifier :
type definition for an Object Identifier data value.
.TP
.B OM_private_object
.IR Private\0Object :
type definition for a handle to an object in an implementation-defined, or private, representation.
.TP
.B OM_public_object
.IR Public\0Object :
type definition for a defined representation of an object that can be directly interrogated by a
programmer.
.TP
.B OM_real
.IR Real :
type definition for a Real data value.
.TP
.B OM_return_code
.IR Return\0Code :
type definition for a value returned from all OM functions indicating either that the function
succeeded or why it failed.
.TP
.B OM_string
.IR String :
type definition for a data value of 'String' syntax.
.TP
.B OM_syntax
.IR Syntax :
type definition for identifying a syntax type.
.TP
.B OM_type
.IR Type :
type definition for an OM attribute type.
.TP
.B OM_type_list
.IR Type\0List :
type definition for a sequence of OM attribute types.
.TP
.B OM_value
.IR Value :
type definition for representing a data value.
.TP
.B OM_value_length
.IR Value\0Length :
type definition for the length of value.
.TP
.B OM_value_position
.IR Value\0Position :
type definition for designating a particular location within a String data value.
.TP
.B OM_workspace
.IR Workspace :
type definition for identifying an application-specific API that implements
OM, such as directory or message handling.
.TE
.\"
.\"
.SH FUNCTIONS
.TP
.BR om_copy ()
.IR Copy ():
This function creates an independent copy of an existing private object, and all of its sub-objects.
The copy is placed in the workspace specified by the client.
.TP
.BR om_copy_value ()
.IR Copy-Value ():
This function replaces an existing attribute value or inserts a new value in one private object with a
copy of an existing attribute value found in another.  Both values must be strings.
.TP
.BR om_create ()
.IR Create ():
This function creates a new private object that is an instance of a particular class.  The object may
be initialized with the attribute values specified as initial in the class' definition.
.sp
The service need not permit the client to explicitly create instances of all classes, but rather only
those indicated, by a package's definition, as having this property.
.TP
.BR om_decode ()
.IR Decode ():
This function creates a new private object that is an exact, but independent, copy of the object that
an existing private object encodes.  The encoding identifies the class of the existing object and the
rules used to encode it.  The allowed rules include, but are not limited to, the BER
and the canonical BER.
.[
x690
.]
.sp
The service need not permit the client to decode instances of all classes, but rather only those
indicated, by a package's definition, as having this property.
.TP
.BR om_delete ()
.IR Delete ():
This function deletes a service-generated public object, or makes a private object inaccessible.
.TP
.BR om_encode ()
.IR Encode ():
This function creates a new private object that encodes an existing private object for conveyance
between workspaces, transport via a network, or storage in a file.  The client identifies the encoding
rules that the service is to follow.  The allowed rules include, but are not limited to, the BER and
the canonical BER.
.[
x690
.]
.sp
The service need not permit the client to encode instances of all classes, but rather only those
indicated, by a package's definition, as having this property.
.TP
.BR om_get ()
.IR Get ():
This function creates a new public object that is an exact, but independent, copy of an existing
private object.  The client may request certain exclusions, each of which reduces the copy to a
portion of the original.  The client may also request that values are converted from one syntax to
another before they are returned.
.sp
The copy may exclude attributes of other than specified types, values at other than specified
positions within an attribute, the values of multi-valued attributes, copies of (not handles for)
sub-objects, or all attribute values (revealing only an attribute's presence).
.TP
.BR om_instance ()
.IR Instance ():
This function determines whether an object is an instance of a particular class.  The client can

NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.


src/lib/xom.c
/*****************************************************************************

 @(#) File: src/lib/xom.c

 -----------------------------------------------------------------------------

 Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 *****************************************************************************/

static char const ident[] = "src/lib/xom.c (" PACKAGE_ENVR ") " PACKAGE_DATE;

/* This file can be processed with doxygen(1). */

/** @weakgroup xom OpenSS7 XOM Library
  * @{ */

/** @file
  * OpenSS7 X/Open Object Model (XOM) implementation file. */

/**
  * This manual contains documentation of the OpenSS7 XOM Library functions that
  * are generated automatically from the source code with doxygen.  This
  * documentation is intended to be used for maintainers of the OpenSS7 XOM
  * Library and is not intended for users of the OpenSS7 XOM Library.  Users
  * should consult the documentation found in the user manual pages beginning
  * with xom(3).
  */

#include <xom.h>

#undef om_copy

/** @brief
  * @param original The original XOM object, which remains accessible.
  * @param workspace The workspace in which the copy is to be created.  The
  *	original's class shall be in a package associated with this workspace.
  * @param copy A pointer to the object to be returned.  The returned object
  *	will only be accessible if the call is successful.
  * @version XOM_1.0
  * @par Alias:
  * This symbol is an implementation of om_copy().
  */
OM_return_code
__xom_om_copy(const OM_private_object original, const OM_workspace workspace,
	      OM_private_object * copy)
{
	if (unlikely(workspace == NULL))
		return (OM_NO_SUCH_WORKSPACE);
	if (unlikely(workspace->functions->omp_copy == NULL))
		return (OM_FUNCTION_DECLINED);
	return (*workspace->functions->omp_copy)
	    (original, workspace, copy);
}

/** @fn OM_return_code om_copy(const OM_private_object original, const OM_workspace workspace, OM_private_object * copy)
  * @param original The original XOM object, which remains accessible.
  * @param workspace The workspace in which the copy is to be created.  The
  *	original's class shall be in a package associated with this workspace.
  * @param copy A pointer to the object to be returned.  The returned object
  *	will only be accessible if the call is successful.
  * @version XOM_1.0
  * @par Alias:
  * This symbol is a strong alias of __xom_om_copy().
  *
  * This function creates a new private object, the @copy, that is an exact, but
  * independent copy of an existing private object, the @original.  The function
  * is recursive in that copying the original also copies its subobjects.
  *
  * Upon successful completion, the zero value of [%OM_SUCCESS] is returned and
  * the new copy of the original is returned in the object pointed to by @copy.
  * Otherwise, a non-zero diagnostic error value is returned and @copy is
  * unaffected.
  *
  * When om_copy() fails, it returns a non-zero error code, as follows:
  *
  * [%OM_NO_SUCH_WORKSPACE]
  *	The @workspace argument is %NULL.
  * [%OM_FUNCTION_DECLINED]
  *	The @workspace does no support the om_copy() function.
  */
__asm__(".symver __xom_om_copy,om_copy@@XDS_1.0");

#undef om_copy_value

/** @brief
  * @param source
  * @param source_type
  * @param source_value_position
  * @param destination
  * @param destination_type
  * @param destination_value_position
  * @version XOM_1.0
  * @par Alias:
  * This symbol is an implementation of om_copy_value().
  */
OM_return_code
__xom_om_copy_value(const OM_private_object source, const OM_type source_type,
		    const OM_value_position source_value_position,
		    OM_private_object destination, OM_type destination_type,
		    OM_value_position destination_value_position)
{
	OMP_workspace workspace;

	if (unlikely(destination == NULL))
		return (OM_NO_SUCH_OBJECT);
	if (unlikely(!(destination->syntax & OM_S_PRIVATE)))
		return (OM_NOT_PRIVATE);
	workspace = (OM_workspace) ((destination - 1)->value.string.elements);
	if (unlikely(workspace->functions->omp_copy_value == NULL))
		return (OM_FUNCTION_DECLINED);
	return (*workspace->functions->omp_copy_value)
	    (source, source_type, source_value_position, destination, destination_type,
	     destination_value_position);
}

/** @fn OM_return_code om_copy_value(const OM_private_object source, const OM_type source_type, const OM_value_position source_value_position, OM_private_object destination, OM_type destination_type, OM_value_position destination_value_position)
  * @param source
  * @param source_type
  * @param source_value_position
  * @param destination
  * @param destination_type
  * @param destination_value_position
  * @version XOM_1.0
  * @par Alias:
  * This symbol is a strong alias of __xom_om_copy_value().
  */
__asm__(".symver __xom_om_copy_value,om_copy_value@@XDS_1.0");

#undef om_create

/** @brief
  * @param class
  * @param initialize
  * @param workspace
  * @param object
  * @version XOM_1.0
  * @par Alias:
  * This symbol is an implementation of om_create().
  */
OM_return_code
__xom_om_create(const OM_object_identifier class, const OM_boolean initialize,
		const OM_workspace workspace, OM_private_object * object)
{
	if (unlikely(workspace == NULL))
		return (OM_NO_SUCH_WORKSPACE);
	if (unlikely(workspace->functions->omp_create == NULL))
		return (OM_FUNCTION_DECLINED);
	return (*workspace->functions->omp_create)
	    (class, initialize, workspace, object);
}

/** @fn OM_return_code om_create(const OM_object_identifier class, const OM_boolean initialize, const OM_workspace workspace, OM_private_object * object)
  * @param class
  * @param initialize
  * @param workspace
  * @param object
  * @version XOM_1.0
  * @par Alias:
  * This symbol is a strong alias of __xom_om_create().
  */
__asm__(".symver __xom_om_create,om_create@@XDS_1.0");

#undef om_decode

/** @brief
  * @param encoding
  * @param original
  * @version XOM_1.0
  * @par Alias:
  * This symbol is an implementation of om_decode().
  */
OM_return_code
__xom_om_decode(const OM_private_object encoding, OM_private_object * original)
{
	OMP_workspace workspace;

	if (unlikely(original == NULL))
		return (OM_NO_SUCH_OBJECT);
	if (unlikely(!(original->syntax & OM_S_PRIVATE)))
		return (OM_NOT_PRIVATE);
	workspace = (OM_workspace) ((original - 1)->value.string.elements);
	if (unlikely(workspace->functions->omp_decode == NULL))
		return (OM_FUNCTION_DECLINED);
	return (*workspace->functions->omp_decode)
	    (encoding, original);
}

/** @fn OM_return_code om_decode(const OM_private_object encoding, OM_private_object * original)
  * @param encoding
  * @param original
  * @version XOM_1.0
  * @par Alias:
  * This symbol is a strong alias of __xom_om_decode().
  */
__asm__(".symver __xom_om_decode,om_decode@@XDS_1.0");

#undef om_delete

/** @brief
  * @param subject
  * @version XOM_1.0
  * @par Alias:
  * This symbol is an implementation of om_delete().
  */
OM_return_code
__xom_om_delete(OM_object subject)
{
	OMP_workspace workspace;

	if (unlikely(subject == NULL))
		return (OM_NO_SUCH_OBJECT);
	if (unlikely(!(subject->syntax & OM_S_SERVICE_GENERATED)))
		return (OM_NOT_THE_SERVICES);
	workspace = (OM_workspace) ((subject - 1)->value.string.elements);
	if (unlikely(workspace->functions->omp_delete == NULL))
		return (OM_FUNCTION_DECLINED);
	return (*workspace->functions->omp_delete)
	    (subject);
}

/** @fn OM_return_code om_delete(OM_object subject)
  * @param subject
  * @version XOM_1.0
  * @par Alias:
  * This symbol is a strong alias of __xom_om_delete().
  */
__asm__(".symver __xom_om_delete,om_delete@@XDS_1.0");

#undef om_encode

/** @brief
  * @param original
  * @param rules
  * @param encoding
  * @version XOM_1.0
  * @par Alias:
  * This symbol is an implementation of om_encode().
  */
OM_return_code
__xom_om_encode(const OM_private_object original, const OM_object_identifier rules,
		OM_private_object * encoding)
{
	OMP_workspace workspace;

	if (unlikely(original == NULL))
		return (OM_NO_SUCH_OBJECT);
	if (unlikely(!(original->syntax & OM_S_PRIVATE)))
		return (OM_NOT_PRIVATE);
	workspace = (OM_workspace) ((original - 1)->value.string.elements);
	if (unlikely(workspace->functions->omp_encode == NULL))
		return (OM_FUNCTION_DECLINED);
	return (*workspace->functions->omp_encode)
	    (original, rules, encoding);
}

/** @fn OM_return_code om_encode(const OM_private_object original, const OM_object_identifier rules, OM_private_object * encoding)
  * @param original
  * @param rules
  * @param encoding
  * @version XOM_1.0
  * @par Alias:
  * This symbol is a strong alias of __xom_om_encode().
  */
__asm__(".symver __xom_om_encode,om_encode@@XDS_1.0");

#undef om_get

/** @brief
  * @param original
  * @param exclusions
  * @param included_types
  * @param local_strings
  * @param initial_value
  * @param limiting_values
  * @param copy
  * @param total_number
  * @version XOM_1.0
  * @par Alias:
  * This symbol is an implementation of om_get().
  */
OM_return_code
__xom_om_get(const OM_private_object original, const OM_exclusions exclusions,
	     const OM_type_list included_types, const OM_boolean local_strings,
	     const OM_value_position initial_value, const OM_value_position limiting_value,
	     OM_public_object * copy, OM_value_position * total_number)
{
	OMP_workspace workspace;

	if (unlikely(original == NULL))
		return (OM_NO_SUCH_OBJECT);
	if (unlikely(!(original->syntax & OM_S_PRIVATE)))
		return (OM_NOT_PRIVATE);
	workspace = (OM_workspace) ((original - 1)->value.string.elements);
	if (unlikely(workspace->functions->omp_get == NULL))
		return (OM_FUNCTION_DECLINED);
	return (*workspace->functions->omp_get)
	    (original, exclusions, included_types, local_strings, initial_value, limiting_value,
	     copy, total_number);
}

/** @fn OM_return_code om_get(const OM_private_object original, const OM_exclusions exclusions, const OM_type_list included_types, const OM_boolean local_strings, const OM_value_position initial_value, const OM_value_position limiting_value, OM_public_object * copy, OM_value_position * total_number)
  * @param original
  * @param exclusions
  * @param included_types
  * @param local_strings
  * @param initial_value
  * @param limiting_values
  * @param copy
  * @param total_number
  * @version XOM_1.0
  * @par Alias:
  * This symbol is a strong alias of __xom_om_get().
  */
__asm__(".symver __xom_om_get,om_get@@XDS_1.0");

#undef om_instance

/** @brief
  * @param subject
  * @param class
  * @param instance
  * @version XOM_1.0
  * @par Alias:
  * This symbol is an implementation of om_instance().
  */
OM_return_code
__xom_om_instance(const OM_object subject, const OM_object_identifier class, OM_boolean * instance)
{
	OMP_workspace workspace;

	if (unlikely(subject == NULL))
		return (OM_NO_SUCH_OBJECT);
	if (unlikely(!(subject->syntax & OM_S_SERVICE_GENERATED)))
		return (OM_NOT_THE_SERVICES);
	workspace = (OM_workspace) ((subject - 1)->value.string.elements);
	if (unlikely(workspace->functions->omp_instance == NULL))
		return (OM_FUNCTION_DECLINED);
	return (*workspace->functions->omp_instance)
	    (subject, class, instance);
}

/** @fn OM_return_code om_instance(const OM_object subject, const OM_object_identifier class, OM_boolean * instance)
  * @param subject
  * @param class
  * @param instance
  * @version XOM_1.0
  * @par Alias:
  * This symbol is a strong alias of __xom_om_instance().
  */
__asm__(".symver __xom_om_instance,om_instance@@XDS_1.0");

#undef om_put

/** @brief
  * @param destination
  * @param modification
  * @param source
  * @param included_types
  * @param initial_value
  * @param limiting_value
  * @version XOM_1.0
  * @par Alias:
  * This symbol is an implementation of om_put().
  */
OM_return_code
__xom_om_put(OM_private_object destination, const OM_modification modification,
	     const OM_object source, const OM_type_list included_types,
	     const OM_value_position initial_value, const OM_value_position limiting_value)
{
	OMP_workspace workspace;

	if (unlikely(destination == NULL))
		return (OM_NO_SUCH_OBJECT);
	if (unlikely(!(destination->syntax & OM_S_PRIVATE)))
		return (OM_NOT_PRIVATE);
	workspace = (OM_workspace) ((destination - 1)->value.string.elements);
	if (unlikely(workspace->functions->omp_put == NULL))
		return (OM_FUNCTION_DECLINED);
	return (*workspace->functions->omp_put)
	    (destination, modification, source, included_types, initial_value, limiting_value);
}

/** @fn OM_return_code om_put(OM_private_object destination, const OM_modification modification, const OM_object source, const OM_type_list included_types, const OM_value_position initial_value, const OM_value_position limiting_value)
  * @param destination
  * @param modification
  * @param source
  * @param included_types
  * @param initial_value
  * @param limiting_value
  * @version XOM_1.0
  * @par Alias:
  * This symbol is a strong alias of __xom_om_put().
  */
__asm__(".symver __xom_om_put,om_put@@XDS_1.0");

#undef om_read

/** @brief
  * @param subject
  * @param type
  * @param value_position
  * @param local_string
  * @param string_offset
  * @param elements
  * @version XOM_1.0
  * @par Alias:
  * This symbol is an implementation of om_read().
  */
OM_return_code
__xom_om_read(const OM_private_object subject, const OM_type type,
	      const OM_value_position value_position, const OM_boolean local_string,
	      const OM_string_length * string_offset, OM_string * elements)
{
	if (unlikely(subject == NULL))
		return (OM_NO_SUCH_OBJECT);
	if (unlikely(!(subject->syntax & OM_S_PRIVATE)))
		return (OM_NOT_PRIVATE);
	if (unlikely(workspace->functions->omp_read == NULL))
		return (OM_FUNCTION_DECLINED);
	return (*OM_WORKSPACE(subject)->functions->omp_read)
	    (subject, type, value_position, local_string, string_offset, elements);
}

/** @fn OM_return_code om_read(const OM_private_object subject, const OM_type type, const OM_value_position value_position, const OM_boolean local_string, const OM_string_length * string_offset, OM_string * elements)
  * @param subject
  * @param type
  * @param value_position
  * @param local_string
  * @param string_offset
  * @param elements
  * @version XOM_1.0
  * @par Alias:
  * This symbol is a strong alias of __xom_om_read().
  */
__asm__(".symver __xom_om_read,om_read@@XDS_1.0");

#undef om_remove

/** @brief
  * @param subject
  * @param type
  * @param initial_value
  * @param limiting_value
  * @version XOM_1.0
  * @par Alias:
  * This symbol is an implementation of om_remove().
  */
OM_return_code
__xom_om_remove(const OM_private_object subject, const OM_type type,
		const OM_value_position initial_value, const OM_value_position limiting_value)
{
	OMP_workspace workspace;

	if (unlikely(subject == NULL))
		return (OM_NO_SUCH_OBJECT);
	if (unlikely(!(subject->syntax & OM_S_PRIVATE)))
		return (OM_NOT_PRIVATE);
	workspace = (OM_workspace) ((subject - 1)->value.string.elements);
	if (unlikely(workspace->functions->omp_remove == NULL))
		return (OM_FUNCTION_DECLINED);
	return (*workspace->functions->omp_remove)
	    (subject, type, initial_value, limiting_value);
}

/** @fn OM_return_code om_remove(const OM_private_object subject, const OM_type type, const OM_value_position initial_value, const OM_value_position limiting_value)
  * @param subject
  * @param type
  * @param initial_value
  * @param limiting_value
  * @version XOM_1.0
  * @par Alias:
  * This symbol is a strong alias of __xom_om_remove().
  */
__asm__(".symver __xom_om_remove,om_remove@@XDS_1.0");

#undef om_write

/** @brief
  * @param subject
  * @param type
  * @param value_position
  * @param syntax
  * @param string_offset
  * @param elements
  * @version XOM_1.0
  * @par Alias:
  * This symbol is an implementation of om_write().
  */
OM_return_code
__xom_om_write(const OM_private_object subject, const OM_type type,
	       const OM_value_position value_position, OM_syntax syntax,
	       const OM_string_length * string_offset, OM_string elements)
{
	OMP_workspace workspace;

	if (unlikely(subject == NULL))
		return (OM_NO_SUCH_OBJECT);
	if (unlikely(!(subject->syntax & OM_S_PRIVATE)))
		return (OM_NOT_PRIVATE);
	workspace = (OM_workspace) ((subject - 1)->value.string.elements);
	if (unlikely(workspace->functions->omp_write == NULL))
		return (OM_FUNCTION_DECLINED);
	return (*workspace->functions->omp_write)
	    (subject, type, value_position, syntax, string_offset, elements);
}

/** @fn OM_return_code om_write(const OM_private_object subject, const OM_type type, const OM_value_position value_position, OM_syntax syntax, const OM_string_length * string_offset, OM_string elements)
  * @param subject
  * @param type
  * @param value_position
  * @param syntax
  * @param string_offset
  * @param elements
  * @version XOM_1.0
  * @par Alias:
  * This symbol is a strong alias of __xom_om_write().
  */
__asm__(".symver __xom_om_write,om_write@@XDS_1.0");

/** @} */

// vim: com=srO\:/**,mb\:*,ex\:*/,srO\:/*,mb\:*,ex\:*/,b\:TRANS
